### ISSUE TITLE
[QUESTION]:  how to split dataset without validation set

### ISSUE THREAD

**Issue by zzzqiii** [chemprop/chemprop#396]
I have some questions about Train/Validation/Test Splits and esemble model. 
1. When using `chemprop_train`, I want to split data into training set and test set and train model on all training set then see the performance of the trained model on test set. I tried to set `--split_sizes 0.8 0 0.2` but received an error that `Provided split size for validation split must be nonzero. Received split size 0.0`.  I'm not sure what the purpose of the validation set in the process of training. Is it used for early stopping and chosing the best epoch? Will validation set data be used for final training? If not, how could I include the validation set into training set? Because my dataset is small, I want the training set to have as much data as possible. 

2. According to the documentation of `Hyperparameter Optimization`, I manually hold out an external test set before hyperoptimization and assign `separate_test_path` when using `chemprop_hyperopt`. However, it seems to choose best hypermeters based on the results of test set rather than validation set which is different with sklearn crossvalidation (https://scikit-learn.org/stable/_images/grid_search_cross_validation.png). I am not very clear about chemprop pipeline of hyperparameter optimization and training. Please tell me the proper usage if my understanding is wrong.

3. I wonder how the esemble models were obtained in chemprop, by training model on different seed of splitting dataset? In my previous practice, I would split dataset set into two part: 80% data was used to construct the model and 20% data was used to test the model performance. The constructing data were randomly divided into training set and validation set for conducting cross-validation. After the cross-validation was complete, the hyperparameters that performed best on the validation set would be retained and used to train on all constructing data. For conducting the ensembling, this procedure would be repeated multiple times with each time performed on a different random train-validation split of the constructing data and multiple models would be retained as esemble models. I tried to use this process but met the two problems mentioned above. Please let me know if any step in the process is not correct. Could you give me some advice? Thanks in advance!

**Comment by cjmcgill** [comment]
Splitting and data pipelining in chemprop doesn't have the builtin complexity that it does in sklearn. You can get to some of the intended behaviors, but it can become quite complex and manual.

1. Yes, the validation set is used for early stopping. The validation set is excluded from the training set. The early stopping is very helpful for your training as neural networks tend to be very overfitting-prone, especially when you have few data. It's a necessary cost; it's worth it. Unfortunately, this does mean that the kinds of splitting and crossvalidation done in sklearn are more complicated to figure out, and not something pipelined by Chemprop, though you could potentially build models to your specification more manually.
2. Hyperparameter optimization doesn't do any special pipeline data handling. It takes in the data and passes it along to individual chemprop training instances which use the data like normal, as train/val/test, with no manipulation . As you say, the result that it picks as the best are the ones that do the best on the test set in the individual trials. If you want to keep your test set un-leaked, then you need to manipulate the data you feed in to exclude your test data. That could mean taking out your test data out and supplying all the rest to --data_path. Or it could mean splitting the data ahead of time and feeding in the training set to --data_path, the validation set to --separate_val_path, and the validation set again to --separate_test_path (that way it makes the decision based on the validation set, but limits your ability to do crossvalidation inside your hyperparameter optimization).
3. Chemprop doesn't have the same nested hyperparameter optimization pipeline that sklearn does. You couldn't do this all in one step with the way Chemprop is currently written. Here's what I can recommend to do I think what you are meaning to accomplish. Do hyperparameter optimization where you've excluded your test set of interest and use some number of cross-validation folds. Then you can use the best hyperparameters indicated there on a subsequent training where you supply your test set of interest as the test set and use crossvalidation to scramble the train/val between folds. You can then make a prediction on your test set using the cross-validation models.

More information:

- You must provide a validation split size. But you can actually run without a test split for training on all the data. You just don't get back a test metric score at the end.
- If you do a prediction on an ensemble of models, those models must share the same hyperparameters in Chemprop. If you really want to ensemble together different model types, then you'd need to make the predictions with each and average them manually.
- There are two kinds of multimodel trainings that chemprop can do. It can do cross-validation where the splitting gets scrambled between folds by using the argument `--num_folds n` (includes changing the test set). If you want for the cross-validation to use a strict crossvalidation scheme where data appears in the test set for exactly one fold, you have to use the argument `--split_type cv`. The second kind of multimodel is ensembling, where the splitting is the same but the random seed initiation is different, set using the argument `--ensemble_size n`. You can do both of these at the same time to have a crossvalidation of ensembles.
- You can supply specific splits for the validation or test sets using `--separate_val_path` and `--separate_test_path`. You can make a set of models where the train/val get scrambled and the test set is kept constant by supplying a `--separate_test_path` and then setting `--num_folds n`. Or you can train individual models independently by manipulating `--seed n` to change the splitting and `--pytorch_seed n` to change the random initialization.
- If you supply val or test splits separately, you don't need to give a split fraction for them.

**Comment by zzzqiii** [comment]
Thanks @cjmcgill ! it helps a lot! I'll give it a try as you suggested.

**Comment by kevingreenman** [comment]
I'm closing this issue for now. Feel free to reopen it if you have more questions about this.
